operator/src/main/java/oracle/kubernetes/weblogic/domain/model/Domain.java

```diff
@@ -628,10 +653,65 @@ public boolean equals(Object other) {
       verifyNoAlternateSecretNamespaceSpecified();
       addMissingModelConfigMap(kubernetesResources);
       verifyIstioExposingDefaultChannel();
+      verifyIntrospectorJobName();
 
       return failures;
     }
 
+    private void verifyIntrospectorJobName() {
+      // K8S adds a 5 character suffix to an introspector job name
+      if (LegalNames.toJobIntrospectorName(getDomainUid()).length()
+          > LegalNames.LEGAL_DNS_LABEL_NAME_MAX_LENGTH - 5) {
+        failures.add(DomainValidationMessages.exceedMaxIntrospectorJobName(getDomainUid()));
+      }
+    }
+
+    private void verifyGeneratedResourceNames(WlsDomainConfig wlsDomainConfig) {
+      checkGeneratedServerServiceName(wlsDomainConfig.getAdminServerName());
+      if (isExternalServiceConfigured(getSpec())) {
+        checkGeneratedExternalServiceName(wlsDomainConfig.getAdminServerName());
+      }
+
+      // domain level serverConfigs do not contain servers in dynamic clusters
+      wlsDomainConfig.getServerConfigs()
+          .values()
+          .stream()
+          .map(WlsServerConfig::getName)
+          .forEach(this::checkGeneratedServerServiceName);
+      wlsDomainConfig.getClusterConfigs()
+          .values()
+          .iterator()
+          .forEachRemaining(wlsClusterConfig
+              // serverConfigs contains configured and dynamic servers in the cluster
+              -> wlsClusterConfig.getServerConfigs().forEach(wlsServerConfig
+                  -> this.checkGeneratedServerServiceName(wlsServerConfig.getName())));
+      wlsDomainConfig.getClusterConfigs()
+          .values()
+          .iterator()
+          .forEachRemaining(wlsClusterConfig -> this.checkGeneratedClusterServiceName(wlsClusterConfig.getName()));
+    }
+
+    private void checkGeneratedExternalServiceName(String adminServerName) {
+      if (LegalNames.toExternalServiceName(getDomainUid(), adminServerName).length()
+          > LegalNames.LEGAL_DNS_LABEL_NAME_MAX_LENGTH) {
+        failures.add(DomainValidationMessages.exceedMaxExternalServiceName(getDomainUid(), adminServerName));
+      }
+    }
+
+    private void checkGeneratedServerServiceName(String serverName) {
+      if (LegalNames.toServerServiceName(getDomainUid(), serverName).length()
+          > LegalNames.LEGAL_DNS_LABEL_NAME_MAX_LENGTH) {
+        failures.add(DomainValidationMessages.exceedMaxServerServiceName(getDomainUid(), serverName));
+      }
+    }
+
+    private void checkGeneratedClusterServiceName(String clusterName) {
+      if (LegalNames.toClusterServiceName(getDomainUid(), clusterName).length()
+          > LegalNames.LEGAL_DNS_LABEL_NAME_MAX_LENGTH) {
+        failures.add(DomainValidationMessages.exceedMaxClusterServiceName(getDomainUid(), clusterName));
```

Comment:
This comment isn't quite in the right place - it belongs on the service name check for server service names within a cluster.

Reply:
I assume when that happens, the introspector will be rerun, and the validation of server service name happens after the introspector is completed, so we should be able to catch this and fail the domain update then with a clear validation error message in the domain resource.

Reply:
In my opinion, there's a risk that this could happen:
- All testing passes at size 5.
- Cluster goes into production and works fine.
- Cluster size is increased from 5 to 15 and fails (perhaps in testing).
- Intrusive post-production changes needed to accomodate the change - including the potential to lose data due to server name changes (file stores and database stores are named after server names).

Reply:
This is a valid and good concern. We could reserve 2 more characters in server service name check to handle this. Alternatively, can customer shortens the domainUID to avoid the issues that you are concerned? DomainUID is only a K8S side artifact that does not have anything to do with the webLogic server side.

Reply:
I'm fine with reducing the maximum domainUID size by approximately 2 characters to account for dynamic clusters that grow to over 100 members. My hope is that there won't be any more need for set-asides like this as we want the maximum to stay subjectively "reasonable."

Reply:
Cool. The key is to change the max size dynamically so that an increased cluster size won't cause trouble.  So decrease the current max by 2 for cluster sizes 1-9, and decrease by 1 for sizes 10-99...

Reply:
I agree with where this is going. 

With the changes in this PR, we enforce the limits at two different levels: 1) we specifies the maximum domainUID size in the domain spec, when this limit is exceeded, helm install will fail. We have similar restrictions to the new suffix values at the helm level; 2) the operator checks the resultant names that take into account of not only the domainUID, but also WebLogic server/cluster names as well as the suffixes. 

I think the reasonable thing to do is to keep the maximum domainUID size (in the spec schema) as high as possible, and leave the flexibility to the choices of suffix values and WebLogic configurations. In order to achieve this, I started to think that we need to allow a custom infix for cluster service names. Once we do that, custom would have almost a full control over the length of the resource names. We could keep the maximum domainUID size to 58 (63 minus the 5 characters that the K8S adds to the job name), instead of 45 (to account for the default suffix plus 5) right now,  as a general limit. Customers have the option to have short or even empty suffixes if needed to leave rooms for other configuration names. The checks in domain validation eventually makes sure that the resultant names are valid although the domainUID size limit in the spec would relatively higher than we current have in this PR.

Reply:
I'm fine with adding support for configuring the cluster infix - it's not critical IMO, but OTOH, now's a good time since 'the patient is open'. 

As for increasing the spec schema max, I don't have a strong opinion but prefer leaving it at 45. That matches the old effective maximum, is relatively 'roomy', and itself only leaves a relatively minimal 23 characters for cluster or server names plus infix decorations.

As for implementing a dynamic max size for clustered server names based on cluster size, an additional consideration is backward compatibility for small clusters that are already at their max. Perhaps it's OK to introduce small backward compatibility issues like this, but, if not, then we could offer configurables to switch back to old behavior: for example 'maxServerServiceNameSizeSmallClusters' and 'maxServerServiceNameSizeMediumClusters'.

If there's time available, it'd be nice if postfix and infix values could be configured on a per domain basis as overrides for the current per operator settings. This would allow a problem domain to be handled without affecting other domains that might already be deployed - and could be especially helpful in production.

Reply:
I also think it is a good time to add customizable cluster infix after all the discussions.
The old effective maximum is 36 (not 45), because we have a longer introspector job name suffix.
I am inclining to not introducing more knobs for backwards compatibility for this. But if we really want to, I marginally prefer adding a boolean to enable/disable the additional calculation for extending cluster size. 
Per domain suffix configuration is nice and more flexible for sure, it time permits.

Reply:
I'm fine with a boolean assuming we choose to worry about backward compat in this area in the first place. I think the two of us have reached a basic agreement on all points.

Reply:
Added code to adjust the upper limit based on the current cluster size with a new flag clusterSizePaddingValidationEnabled to enable additional padding validation.